File: traceharbor-resource-canary/traceharbor-resource-canary-android/src/main/java/com/kernelflux/traceharbor/resource/processor/AutoDumpProcessor.java
package com.kernelflux.traceharbor.resource.processor;

import com.kernelflux.traceharbor.resource.analyzer.model.DestroyedActivityInfo;
import com.kernelflux.traceharbor.resource.analyzer.model.HeapDump;
import com.kernelflux.traceharbor.resource.watcher.ActivityRefWatcher;
import com.kernelflux.traceharbor.util.TraceHarborLog;

import java.io.File;

/**
 * Created by Yves on 2021/3/4
 */
public class AutoDumpProcessor extends BaseLeakProcessor {

    private static final String TAG = "TraceHarbor.LeakProcessor.AutoDump";

    public AutoDumpProcessor(ActivityRefWatcher watcher) {
        super(watcher);
    }

    @Override
    public boolean process(DestroyedActivityInfo destroyedActivityInfo) {
        final File hprofFile = getHeapDumper().dumpHeap(true);
        if (hprofFile != null) {
            getWatcher().markPublished(destroyedActivityInfo.mActivityName);
            getWatcher().triggerGc();
            final HeapDump heapDump = new HeapDump(hprofFile, destroyedActivityInfo.mKey, destroyedActivityInfo.mActivityName);
            getHeapDumpHandler().process(heapDump);
        } else {
            TraceHarborLog.i(TAG, "heap dump for further analyzing activity with key [%s] was failed, just ignore.",
                    destroyedActivityInfo.mKey);
        }
        return true;
    }
}

File: traceharbor-resource-canary/traceharbor-resource-canary-android/src/main/java/com/kernelflux/traceharbor/resource/processor/AutoDumpProcessor.kt
package com.kernelflux.traceharbor.resource.processor

import com.kernelflux.traceharbor.resource.analyzer.model.DestroyedActivityInfo
import com.kernelflux.traceharbor.resource.analyzer.model.HeapDump
import com.kernelflux.traceharbor.resource.watcher.ActivityRefWatcher
import com.kernelflux.traceharbor.util.TraceHarborLog

class AutoDumpProcessor(
    watcher: ActivityRefWatcher,
) : BaseLeakProcessor(watcher) {
    override fun process(destroyedActivityInfo: DestroyedActivityInfo): Boolean {
        val hprofFile = heapDumper.dumpHeap(true)
        if (hprofFile != null) {
            watcher.markPublished(destroyedActivityInfo.mActivityName)
            watcher.triggerGc()
            val heapDump = HeapDump(hprofFile, destroyedActivityInfo.mKey, destroyedActivityInfo.mActivityName)
            heapDumpHandler.process(heapDump)
        } else {
            TraceHarborLog.i(
                TAG,
                "heap dump for further analyzing activity with key [%s] was failed, just ignore.",
                destroyedActivityInfo.mKey,
            )
        }
        return true
    }

    companion object {
        private const val TAG = "TraceHarbor.LeakProcessor.AutoDump"
    }
}

